### PR TITLE
Implement a first version of the computation graph

### PR DESCRIPTION
**Pull request by marcenacp**
This PR allows to convert the Croissant file to a verifiable computation graph.

The "structure graph" is a Python representation of the Croissant file. It describes the structure of the dataset with the following `Nodes`:
- `Field`
- `FileObject`
- `FileSet`
- `Metadata`
- `RecordSet`

The "computation graph" contains the set of instructions to execute to generate the dataset. For example:
image

To be improved in this PR or the next PRs:
- More unit tests (now I can easily write them with the `Node` abstraction, because I can easily create nodes as Python objects).
- Parsing RDF->structure graph. I used `rdflib.Graph()` because it was easy, but I now feel like it complexifies the code. This ended up adding a layer of indirection. Maybe we could remove this layer.
- Add more potential issues (for instance, check that `url` properties are URLs, etc).
- Documentation (writing a design doc on the side).
- Better context for `Issues`. The context could be attached to each node.
- Overload `nx.Graph` to factorize some functions.

### CODE REVIEW

File: datasets/pass/metadata.json [hunk 1]
```diff
@@ -3,10 +3,17 @@
     "@vocab": "https://schema.org/",
     "sc": "https://schema.org/",
     "ml": "http://mlcommons.org/schema/",
-    "includes": {
-      "@id": "ml:includes",
-      "@type": "@id"
-    }
+    "includes": "ml:includes",
```
Comment (pierrot0) [inline datasets/pass/metadata.json:6]: Do we need to define those somewhere?

File: datasets/titanic/metadata.json [hunk 2]
```diff
@@ -3,19 +3,17 @@
       "@vocab": "https://schema.org/",
       "sc": "https://schema.org/",
       "ml": "http://mlcommons.org/schema/",
-      "wd": "https://www.wikidata.org/wiki/Q",
```
Comment (pierrot0) [inline datasets/titanic/metadata.json:6]: Is this deleted intentionally?

**Review by pierrot0 — COMMENTED** [review]
LGTM overall, just a few minor comments.

File: python/format/src/computations.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    entry_nodes = []
+    for node, indegree in graph.in_degree(graph.nodes()):
+        if indegree == 0:
+            entry_nodes.append(node)
+    return entry_nodes
+
+
+def get_structure_graph(
+    issues: Issues, nodes: list[Node]
+) -> tuple[Node, nx.MultiDiGraph]:
```
[the remark this answers is not in the document]
Comment (marcenacp) [reply python/format/src/computations.py]: Done.

File: python/format/src/computations.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                            node.encoding_format == "application/x-tar"
+                            and isinstance(successor, (FileObject, FileSet))
+                            and successor.encoding_format != "application/x-tar"
+                        ):
+                            operation = Operation(name=f"extract:{node.uid}")
+                            operations.add_edge(
+                                last_operation_for_node[node], operation
+                            )
+                            last_operation_for_node[node] = operation
+                elif isinstance(node, (FileObject, FileSet)):
```
[the remark this answers is not in the document]
Comment (marcenacp) [reply python/format/src/computations.py:171]: This is the expected behaviour.

(FileObjects need to be downloaded, FileSets do not.)

File: python/format/src/computations.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+@dataclasses.dataclass(frozen=True)
+class ComputationGraph:
+    """Graph of dependent operations to execute to generate the dataset."""
+
+    issues: Issues
+    graph: nx.MultiDiGraph
+
+    @classmethod
+    def from_nodes(self, issues: Issues, nodes: list[Node]) -> "ComputationGraph":
```
[the remark this answers is not in the document]
Comment (marcenacp) [reply python/format/src/computations.py:104]: Done.

File: datasets/pass/metadata.json (hunk 1 above)
Comment (marcenacp) [reply datasets/pass/metadata.json:6]: For the moment, all constants are defined in python/format/src/constants.py.

`"includes": "ml:includes"` is a context shortcut to map `"includes"` to `"http://mlcommons.org/schema/includes"`.

Did you mean anything different?

File: python/format/src/computations.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        for layer in nx.bfs_layers(graph, entry_node):
+            for node in layer:
+                predecessors = graph.predecessors(node)
+                for predecessor in predecessors:
+                    if predecessor in last_operation_for_node:
+                        last_operation_for_node[node] = last_operation_for_node[
+                            predecessor
+                        ]
+                if isinstance(node, Field):
+                    if node.source and not node.has_sub_fields:
```
[the remark this answers is not in the document]
Comment (marcenacp) [reply python/format/src/computations.py:126]: This function is just a PoC to list operations to perform. Not sure we keep it later as is. That's why operations are strings at this point (e.g., `Operation(name="foo")`).

I would advise to wait for the next PR, when we have a better idea of what operations should look like in practice.

**Review by marcenacp — COMMENTED** [review]
Thanks!

**Review by pierrot0 — APPROVED** [review]

**Pull request merged**